### ISSUE TITLE
Feature request: constrain window to video aspect ratio

### ISSUE THREAD

**Issue by willgfx** [Rafostar/clapper#30]
It would be nice to be able constrain the window to the video's aspect ratio when manually resizing.

**Comment by Rafostar** [comment]
I always wanted to have the resize work like that, but it is not easy to achieve in GTK apps. If I am not mistaken other GTK video players don't have that ability either (`totem` and `celluloid`).

I would prefer to have the video rendering widget merged into `GStreamer` first, cause it might require some changes to get working what you ask for. Now that GTK 4.0 is released (and considered stable API), I will work towards having that updated and merged. Recently being able to toggle playback remotely from the web browser took priority for my "couch-friendly" holiday season :christmas_tree: :santa: :couch_and_lamp:

**Comment by Rafostar** [comment]
A user provided a video with [this demonstration](https://matrix-client.matrix.org/_matrix/media/r0/download/gnome.org/9bb17342d5440ebdf83e00095e286a47ee4ab914).

I think that we can do something similar here for this functionality and make it an opt-in in preferences. So when enabled in Clapper prefs, after user finishes dragging the window to such-and-such size (with black borders), Clapper window automatically resizes itself to best possible (nearest) size for played video, reducing its size by borders amount (with fluid animation, like we have for show/hide bottom panel when clicking PiP button).

**Comment by frandavid100** [comment]
That would be great.

**Comment by endim8** [comment]
This would be really nice. Firefox's PiP window is able to keep a set aspect ratio, maybe looking at that implementation could help?

**Comment by nekohayo** [comment]
Besides the whole "preserve aspect ratio while the user is resizing" idea, I think a good first step would be to have clapper auto-size the window to the correct size and aspect ratio when it is launched directly with a video file (i.e. by Nautilus).

If the window is already in the correct aspect ratio, and auto-sized to fit most needs (i.e. look a the video's resolution vs the screen resolution, and try to fill maybe 80% of the screen if the video's resolution isn't too low) from the start, then it would already do most of the job and the user wouldn't even need to resize in most cases.

This could be done independently as a first step, before a (more complicated) solution is found for constraining.

**Comment by RedAuburn** [comment]
the main usage i would have for this is when the player is in a sort of picture-in-picture mode, like the old version used to have. Only constraining aspect ratio when in PiP mode would work well IMO

**Comment by Rafostar** [comment]
> the player is in picture-in-picture mode, like the old version used to have

We do not have a different/dedicated PiP mode in new UI, but it doesn't mean there is no such functionality. All you have to do is right click anywhere on the video and select always on top option from context menu. This avoids us ugly hacks that we had previously and even stopped working after some GNOME shell update.

Basically, what I mean is that we need a single way for this feature to work the same regardless if always on top is selected or not, since mutter handles that and we aren't aware of it's state here.

**Comment by Rafostar** [comment]
Other problem that we want to avoid, example:

User constrains window size, selects always on top option, puts video in screen corner, then plays few videos in queue with order with aspect ratios: `16/9` -> `4/3` -> `16/9` -> `21:9` -> `16/9`.

What happens to the window then? Does it stay intact? Does it grow or shrink when going to different aspect (also consider how these aspect differs - black bars at different side depending on video). Also if it was at some size initially, will it go back to this initial size after all this aspect changes? When video width is longer will it go outside the screen on its own?

For these reasons I suggested manual "resize to best fit" action with a middle button click on video. Cause the benefit of it is that if you lets say play 7 videos with the same aspect and some very short clip at different aspect, then size would change and it would be hard/impossible to get it back into exact same size and place automatically where it was earlier.

It can be always changed in future, but hey, its much better then nothing, right? :wink:

**Comment by Rafostar** [comment]
> Taking above into account, I suggest assigning mouse middle click to this.

Nobody seems to be against, so I went ahead and implemented just that:

[out.webm](https://github.com/Rafostar/clapper/assets/40623528/e8ca2644-0b37-425d-907d-d7b0e6cd4f56)

(note that PC I was recording this sucks - its recording task that causes slowdown/stutter)

**Comment by sp1ritCS** [comment]
> Nobody seems to be against

well, the downside is that it doesn't work without interaction. Maybe the usecase of having a small clapper window as pip in the corner of the screen, while having an application that captures the mouse (mostly video games) might be slightly impaired.

I doubt it's that really relevant tho.

**Comment by Rafostar** [comment]
> well, the downside is that it doesn't work without interaction.

yeah, but there are multiple things that may go haywire if I try to make it force itself to keep aspect ratio while resizing. GTK does not do this for a reason. Also, I am not closing this issue.

The main use-case for this is PiP like "always on top" usage of Clapper window, yes. On the other side, with this you can middle click on a window, put it in a corner and forget its there. Automatic behavior would each time a video with a different aspect is played cause window to change its size and position on its own (we cannot control window position on Wayland) which seems like it would actually need more interactions to fix its position manually by user.

**Comment by EPOCHvoyager** [comment]
I like the way MPC-HC on Windows and now Showtime do it, in that they resize the video only on opening the file. That largely avoids unnecessary scaling and letterboxing, and doesn't affect PiP usage where it doesn't matter as much. 

Also, when the video's resolution would be too large or matches your desktop, a larger - but clamped, leaving some space around the player - window size would be a nice hint that you might want to go fullscreen. I believe MPC-HC does something like this.
